// @hashgraph/sdk/lib/TransactionFeeSchedule.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var HieroProto = _interopRequireWildcard(require("@hashgraph/proto"));
var _RequestType = _interopRequireDefault(require("./RequestType.cjs"));
var _FeeData = _interopRequireDefault(require("./FeeData.cjs"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
// SPDX-License-Identifier: Apache-2.0

class TransactionFeeSchedule {
  /**
   * @param {object} [props]
   * @param {RequestType} [props.hederaFunctionality]
   * @param {FeeData} [props.feeData]
   * @param {FeeData[]} [props.fees]
   */
  constructor(props = {}) {
    /*
     * A particular transaction or query
     *
     * @type {RequestType}
     */
    this.hederaFunctionality = props.hederaFunctionality;

    /*
     * Resource price coefficients
     *
     * @type {FeeData}
     */
    this.feeData = props.feeData;

    /*
     * Resource price coefficients
     *
     * @type {FeeData[]}
     */
    this.fees = props.fees;
  }

  /**
   * @param {Uint8Array} bytes
   * @returns {TransactionFeeSchedule}
   */
  static fromBytes(bytes) {
    return TransactionFeeSchedule._fromProtobuf(HieroProto.proto.TransactionFeeSchedule.decode(bytes));
  }

  /**
   * @internal
   * @param {HieroProto.proto.ITransactionFeeSchedule} transactionFeeSchedule
   * @returns {TransactionFeeSchedule}
   */
  static _fromProtobuf(transactionFeeSchedule) {
    return new TransactionFeeSchedule({
      hederaFunctionality: transactionFeeSchedule.hederaFunctionality != null ? _RequestType.default._fromCode(transactionFeeSchedule.hederaFunctionality) : undefined,
      feeData: transactionFeeSchedule.feeData != null ? _FeeData.default._fromProtobuf(transactionFeeSchedule.feeData) : undefined,
      fees: transactionFeeSchedule.fees != null ? transactionFeeSchedule.fees.map(fee => _FeeData.default._fromProtobuf(fee)) : undefined
    });
  }

  /**
   * @internal
   * @returns {HieroProto.proto.ITransactionFeeSchedule}
   */
  _toProtobuf() {
    return {
      hederaFunctionality: this.hederaFunctionality != null ? this.hederaFunctionality.valueOf() : undefined,
      feeData: this.feeData != null ? this.feeData._toProtobuf() : undefined,
      fees: this.fees != null ? this.fees.map(fee => fee._toProtobuf()) : undefined
    };
  }

  /**
   * @returns {Uint8Array}
   */
  toBytes() {
    return HieroProto.proto.TransactionFeeSchedule.encode(this._toProtobuf()).finish();
  }
}
exports.default = TransactionFeeSchedule;
