// @hashgraph/sdk/lib/address_book/NodeAddressBook.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _NodeAddress = _interopRequireDefault(require("./NodeAddress.cjs"));
var HieroProto = _interopRequireWildcard(require("@hashgraph/proto"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
// SPDX-License-Identifier: Apache-2.0

/**
 * @typedef {import("./NodeAddress.js").NodeAddressJson} NodeAddressJson
 */

/**
 * @typedef {object} NodeAddressBookJson
 * @property {NodeAddressJson[]} nodeAddresses
 */

/**
 * Represents a collection of node addresses in the Hedera network.
 *
 * The NodeAddressBook contains information about the nodes in the Hedera network,
 * including their network addresses, account IDs, and node IDs. This class is used
 * to manage and access the network's node information.
 */
class NodeAddressBook {
  /**
   * @param {object} props
   * @param {NodeAddress[]} [props.nodeAddresses]
   */
  constructor(props = {}) {
    /**
     * @type {NodeAddress[]}
     */
    this._nodeAddresses = [];
    if (props.nodeAddresses != null) {
      this.setNodeAddresses(props.nodeAddresses);
    }
  }

  /**
   * @returns {NodeAddress[]}
   */
  get nodeAddresses() {
    return this._nodeAddresses;
  }

  /**
   * @param {NodeAddress[]} nodeAddresses
   * @returns {this}
   */
  setNodeAddresses(nodeAddresses) {
    this._nodeAddresses = nodeAddresses;
    return this;
  }

  /**
   * @param {Uint8Array} bytes
   * @returns {NodeAddressBook}
   */
  static fromBytes(bytes) {
    return NodeAddressBook._fromProtobuf(HieroProto.proto.NodeAddressBook.decode(bytes));
  }

  /**
   * @internal
   * @param {HieroProto.proto.INodeAddressBook} nodeAddressBook
   * @returns {NodeAddressBook}
   */
  static _fromProtobuf(nodeAddressBook) {
    return new NodeAddressBook({
      nodeAddresses: nodeAddressBook.nodeAddress != null ? nodeAddressBook.nodeAddress.map(nodeAddress => _NodeAddress.default._fromProtobuf(nodeAddress)) : undefined
    });
  }

  /**
   * @returns {HieroProto.proto.INodeAddressBook}
   */
  _toProtobuf() {
    return {
      nodeAddress: this._nodeAddresses.map(nodeAddress => nodeAddress._toProtobuf())
    };
  }

  /**
   * @returns {string}
   */
  toString() {
    return JSON.stringify(this.toJSON());
  }

  /**
   * @returns {NodeAddressBookJson}
   */
  toJSON() {
    return {
      nodeAddresses: this._nodeAddresses.map(nodeAddress => nodeAddress.toJSON())
    };
  }
  toBytes() {
    return HieroProto.proto.NodeAddressBook.encode(this._toProtobuf()).finish();
  }
}
exports.default = NodeAddressBook;
